pkg/server/commands/listusers/list_users_rpc_test.go

```diff
@@ -2,301 +2,1132 @@ package listusers
+						Type: "user",
+					},
+				},
+			},
+			model: `model
+			schema 1.1
+			type user
+			type group
+				relations
+					define member: [user]
 
+			type document
+				relations
+					define viewer: [ group:*, user:*]`,
+
+			tuples: []*openfgav1.TupleKey{
+				tuple.NewTupleKey("document:1", "viewer", "user:*"),
+				tuple.NewTupleKey("document:1", "viewer", "group:*"),
+			},
+			expectedUsers: []string{"user:*"},
+		},
+		{
+			name: "multiple_possible_wildcards_group_granularity",
+			req: &openfgav1.ListUsersRequest{
+				Object:   &openfgav1.Object{Type: "document", Id: "1"},
+				Relation: "viewer",
+				UserFilters: []*openfgav1.ListUsersFilter{
+					{
+						Type: "group",
+					},
+				},
+			},
+			model: `model
+			schema 1.1
+			type user
+			type group
+				relations
+					define member: [user]
+
+			type document
+				relations
+					define viewer: [ group:*, user:*]`,
+
+			tuples: []*openfgav1.TupleKey{
+				tuple.NewTupleKey("document:1", "viewer", "user:*"),
+				tuple.NewTupleKey("document:1", "viewer", "group:*"),
+			},
+			expectedUsers: []string{"group:*"},
+		},
+	}
+	tests.runListUsersTestCases(t)
+}
+
+func (testCases ListUsersTests) runListUsersTestCases(t *testing.T) {
+	storeID := ulid.Make().String()
+
+	for _, test := range testCases {
 		ds := memory.New()
 		defer ds.Close()
+		model := testutils.MustTransformDSLToProtoWithID(test.model)
 
-		model := &openfgav1.AuthorizationModel{
-			Id:              modelID,
-			TypeDefinitions: test.typeDefinitions,
-			SchemaVersion:   typesystem.SchemaVersion1_1,
-		}
-
-		typesys, err := typesystem.NewAndValidate(context.Background(), model)
-		require.NoError(t, err)
+		t.Run(test.name, func(t *testing.T) {
+			if test.TemporarilySkipReason != "" {
+				t.Skip()
+			}
 
-		err = ds.WriteAuthorizationModel(context.Background(), storeID, model)
-		require.NoError(t, err)
+			typesys, err := typesystem.NewAndValidate(context.Background(), model)
+			require.NoError(t, err)
 
-		if len(test.tuples) > 0 {
-			err = ds.Write(context.Background(), storeID, nil, test.tuples)
+			err = ds.WriteAuthorizationModel(context.Background(), storeID, model)
 			require.NoError(t, err)
-		}
 
-		l := NewListUsersQuery(ds)
+			if len(test.tuples) > 0 {
+				err = ds.Write(context.Background(), storeID, nil, test.tuples)
+				require.NoError(t, err)
+			}
 
-		ctx := typesystem.ContextWithTypesystem(context.Background(), typesys)
+			l := NewListUsersQuery(ds)
 
-		t.Run(test.name, func(t *testing.T) {
+			ctx := typesystem.ContextWithTypesystem(context.Background(), typesys)
+
+			test.req.AuthorizationModelId = model.GetId()
+			test.req.StoreId = storeID
 
 			resp, err := l.ListUsers(ctx, test.req)
 			require.ErrorIs(t, err, test.expectedError)
 
-			ignoredFieldsOpts := protocmp.IgnoreFields(protoadapt.MessageV2Of(&openfgav1.Object{}), "type", "id")
+			actualUsers := resp.GetUsers()
+
+			actualCompare := make([]string, len(actualUsers))
+			for i, u := range resp.GetUsers() {
```

Comment:
Can we please make the conversion of `*openfgav1.User` to `string` a utility of the `testutils` package and delegate to it here?

```go
package testutils

func UserProtobufsToStrings(users *openfgav1.User...) []string {
   ...
}
```

then we have
```go
actual := testutils.UserProtobufsToStrings(resp.GetUsers()...)
require.ElementsMatch(t, actual, test.expectedUsers)
```

Reply:
Agree on the helper, but i would put it in the `tuple` pkg because that has been the standard place for all converters, e.g. https://github.com/openfga/openfga/blob/7bb7fe40ab28610994cfdd75744edb0ae6523d8e/pkg/tuple/tuple.go#L151-L168

Reply:
100% agree with a dedicated conversion utility function. Are we ok with that being a quick follow-up PR? This PR is already quite large and the aforementioned work will also itself require tests.